Rahavaraukset master

rckopteri opened the pull request:
<div align="center">

# Rahavaraukset-master

![image](https://github.com/user-attachments/assets/9294f446-15e0-4ed1-8440-de4008e30077)
![image](https://github.com/user-attachments/assets/15219a97-c0e5-4511-a79a-2eaeaa9bc036)
</div>

## Sopimusmuutokset 2024, lisätään rahavaraukset kustikseen, kuluihin, laskutusyhteenvetoihin, tietokantaan ja kustannusten seurantaan

• Suunnniteltavat rahavaraukset 
• Kustannusten seuranta 
• Tuotekohtainen laskutusyhteeveto 
• Työmaakokous laskutusyhteeveto 
• Kulujen kohdistus 


## Tietokanta: 

> - Rahavarauksiin liittyvät tietokantamuutokset, jotka eivät aiheuta muutostarpeita muualla koodissa tehty erillisesti: [3517](https://github.com/finnishtransportagency/harja/pull/3517)

> Tähän branchiin on kuitenkin jäänyt vielä paljon erityisesti historiaan vaikuttavia tietokantamuutoksia
> Kuten:
>   • kustannusarvioitu_tyo tauluun päivitetään kaikille vanhoille rahavarauksille uuden rahavarausmallin mukainen rahavaraus_id
>   • kulu_kohdistus tauluun tulee selkeämmin tieto, että minkä tyyppinen kulu on. Kuluja on nyt: hankintakulu, rahavaraus, lisätyö ja muu-kulu. Vanhat kulut muokataan näihin tyyppeihin sopiviksi
>   • Nuo kaksi muutosta siis vaikuttavat koko harjan historiaan (ehkä vuodesta 2019, kun MHU-malli on otettu käyttöön) ja ovat täten aika kriittisiä


## Laskutus: 

> - Korvattu 'maksueratyyppi' -> rahavaraus_id:ksi kun haetaan tai filtteröidään kuluja rahavarauksilla (Äkilliset hoitotyöt / vahinkojen korjaukset / kannustinjärjestelmä)
> - Vanhat termit ja muuttujat heitetty kuikkaan, uusi tietomalli korvaa (termit haetaan taulusta)
> - Kaikki rahavaraukset haetaan suoraan kannasta arrayhyn, joka mäpätään Gridiin
> - Näytetään vain urakan rahavaraukset, tietyssä järjestyksessä
> - Refaktoroitu pohja 
> - Lisätty luoja sarake toteutuneet_kustannukset tauluun
> - Poistettu duplikaatti rahavaraukset kannasta (kaikkia tulee olla vain 1)
> - Rahavaraus_id populointi (laitettu ajoon, populoi_rahavaraus_idt)
> - Päällysteiden paikkausten toimenpideinstanssit korjattu ([709](https://extranet.vayla.fi/jira/browse/HARJA-709)) (tehtava, kulu_kohdistus)
> - Korjattu lisätyöt laskutusyhteevedolle 
> - Muut kulut ([609](https://extranet.vayla.fi/jira/browse/HARJA-609)) tavoitehintaiset, ei tavoitehintaiset
> - Työmaakokoukselle uusi Layout, figma speksi (Siivottu 18.9): ([355](https://extranet.vayla.fi/jira/browse/HARJA-355))
> - Tuotekohtaisessa näytetään instanssien alla vain instansseihin liittyvät rahavaraukset 


## Kulujen kohdistus: 

> - Kulun lisäyslomake koodattu uudelleen 
> - Kulun lisäämisessä näytetään myös päätöksen sisältävät kuukaudet, mutta nämä on lukittu eikä kulua voi lisätä. Tähän lisätty infolaatikko sekä lukitusikonit, joka ohjeistaa käyttäjää 
> - Koontilaskun kuukausi ja päivämäärävalintakenttien rajoituksia on muutettu hieman, jotta käyttäjä ei olisi niin lukossa niiden eräpäivän valinnan kanssa
> - Ulkoasua on muutettu rajusti. Nyt lomakkeen pitäisi toimia kohtuullisesti myös puhelimella ja muutenkin olla paremmin responsiivinen

## Kulujen listaus:
> - Listausnäkymään on lisätty hotokausi -alasvetovalikko. Sen toiminta vaikuttaa kaikkiin muihinkin filttereihin sivulla
> - Listausnäkymän aikamäärän valintakentät on tehty uusiksi, jotta niihin voidaan päivittää (lähinnä poistaa) päivämäärä, kun jotakin toista sivun filtteriä käytetään.


## Kustannussuunnitelma: 

> - Kustiksesta on poistettu hankintakulujen toimenpiteiden alta (Liikenneympäristönhoito, MHU Ylläpito) rahavaraukset
> - Rahavaraukset on lisätty omaan karsinaansa ja koodattu ilman pohjalla olevaa "gridiä"
> - Tavoitehinnan ulkopuoliset rahavaraukset on tuotu lomakkeen pohjalta rahavarausosioon ja koodattu samalla tavalla uusiksi ilman "gridiä"
> - Tavoitehintaisille rahavarauksille on lisätty tietokantaan oma uusi osio, joten ne on nyt mahdollista indeksivahvistaa erikseen. (tämä näkyy kustannusten seurannassa)

## Kustannusten seuranta: 

> - Rahavaraukset on irroitettu hankintakulujen alta
> - Rahavarauksilla on nyt oma "pääryhmä", jonka alle ne listataan tietokantaa määritellyssä järjestyksessä.
> - Rahavarausten budjetoitu osioon summat tulevat kustiksesta. Jos toteumia ei ole, niin rahavarauksista ei näytetä kuin 2 tasoa (pääryhmä, rahavaraustaso). Jos kulu puolelta on lisätty toteutumakulua, niin näytetään myös tehtäväryhmä, jolle kulu on lisätty.
> - Kulupuolen muut-kulut tyyppiset kulut näytetään erikeen omana pääryhmänään. Muut kulut voivat olla joko tavoitehintaisia tai ei-taivoitehintaisia. Sen mukaan ne järjestetään sivulla omaan ryhmäänsä.

solita-staskila committed:
Populoi rahavaraus id:t testidataan

solita-staskila committed:
Korjaa testi arvoja

solita-staskila committed:
Korjaa duplikaatti rahavaraukset

solita-staskila committed:
Poista funktion ajo migraatiosta

solita-staskila committed:
Sievennä turhaa iffittelyä

rckopteri committed:
Lisätään rahavaraushakuun tehtävä, tehtäväryhmä ja toimenpide

koska kulut tarvitsevat niitä

rckopteri committed:
Mahdollistetaan rahavarausten tehtäväryhmien hakeminen kuluille

rckopteri committed:
Mahdollista rahavarausten tehtäväryhmien haku

rckopteri committed:
Lisää muu-kululle validoinnit

rckopteri committed:
Alustavat toiminnat muu-kulu tyyppiselle kululle

rckopteri committed:
Aloita hoitovuosikohtaisen alasvetovalikon suunnittelu

rckopteri committed:
Lisää tuki hoitovuoden päätöstyyppisille kuluille

rckopteri committed:
Jalosta kulujen ja budjetointien mäppäystä rahavarauksiksi

rckopteri committed:
Poista turhat kolumnit ja lisää kululle uusi tyyppi

rckopteri committed:
Merge branch 'develop' into HARJA-318-rahavar-kustannusseuranta

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1115__.sql

rckopteri committed:
Poista muutamia virheita clj-kondosta

rckopteri committed:
Lisää Muut rahavaraukset esimerkin vuoksi, jotta sen voi valita kululle ja näyttää kustannusten seurannassa

rckopteri committed:
Poista kulun tyyppi sql hauista

rckopteri committed:
Deklaroi muutama funktio, jotta clj-kondo on tyytyväinen

solita-staskila committed:
Formatointi

rckopteri committed:
Poista turhat kolumnit ja lisää tarvittavat rahavaraukset

rckopteri committed:
Muokkaa kulut uuteen tietomalliin

rckopteri committed:
Päivitä kulujen sql haut uuteen tietomalliin

rckopteri committed:
Päivitä kulupalvelu vastaamaan muuttunutta tietokantatoteutusta

rckopteri committed:
Validoi kululomake ja refaktoroi

rckopteri committed:
Muuta kulujen testi vastaamaan muuttunutta toteutusta

rckopteri committed:
Pieni refaktorointi

rckopteri committed:
Pieni refaktorointi

rckopteri committed:
Korjaa hoitokausivalinnan toimintaa

rckopteri committed:
Merge branch 'refs/heads/develop' into HARJA-318-rahavar-kustannusseuranta

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1116__.sql

rckopteri committed:
Pieni parannus kustannusten seurantaan ja rahavarausten hakuun

rckopteri committed:
Näytä lisätyöt ja muut kulut omina pääryhminään

rckopteri committed:
Korjaa testidataan kulujen tietorakenne muutoksien aiheuttamat ongelmat

rckopteri committed:
Muokkaa kustannusten seurannan testit vastaamaan uutta kulujen tietorakennetta

rckopteri committed:
Poista mahdollinen NPE

rckopteri committed:
Muuta kohdistustyyppi tyypiksi, kuten tietokantakolumni

rckopteri committed:
Muuta testit vastaamaan taas todellisuutta

rckopteri committed:
Päivitä testejä tietorakenteen muuttuessa

rckopteri committed:
Laitetaan ehdolle rahavarausten muutokset

solita-staskila committed:
Merge branch 'HARJA-355' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

solita-staskila committed:
Merge branch 'HARJA-318-rahavar-kustannusseuranta' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

solita-staskila committed:
Lisää konfliktin migraatio

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1119__.sql

rckopteri committed:
Poista johto ja hallintokorvauksista toimenpidetaso (palkat ja toimistokulut)

rckopteri committed:
Lisää muut kulut -  johto ja hallintokorvaukset käsin kirjattuna

rckopteri committed:
Poista kuollutta koodia

rckopteri committed:
UI formatointia

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	src/clj/harja/kyselyt/rahavaraukset.sql
#	tietokanta/src/main/resources/db/migration/V1_1123__.sql

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

rckopteri committed:
Lisää rahavarauksille vahvistamistyyppi tietokantaan

rckopteri committed:
Tavoitehintaisten rahavarausten muokkaus omaan osioon

rckopteri committed:
Ota tavoitehintaiset rahavaraukset osio käyttöön kustannussuunnitelma viewissä

rckopteri committed:
Lisää uusi vahvistustyyppi tavoitehintaisille rahavarauksille

rckopteri committed:
Hae, muokkaa ja laske yhteen tavoitehintaisten rahavarausten tiedot

rckopteri committed:
Mahdollista yhteenvetorivin lisääminen muokkaus-gridille

rckopteri committed:
Lisää uusi endpoint tavoitehintaisille rahavarauksille.

Kaikki on kovakoodattua ja mitään ei oikeasti haeta tai tallenneta kannan kautta.

rckopteri committed:
Formatoi summa ja pakota se positiiviseksi luvuksi

rckopteri committed:
Ota indeksilaskennat mukaan rahavarauksiin

rckopteri committed:
Vähä häkki, mutta näin saadaan muutamalle urakalle rahavaraus.

rckopteri committed:
Estä turhat backend kutsut

rckopteri committed:
HAe ja tallenna rahavaraukset kustiksessa

rckopteri committed:
Poista duplikaatti populointi

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1124__.sql
#	tietokanta/testidata/analytiikka-paallystyskohteet.sql

rckopteri committed:
Nimeä R migraatio isolla kirjaimella

rckopteri committed:
Päivitä migraatiot toimimaan uusilla rahavarausnimillä

rckopteri committed:
Lisää käyttöoikeustesti tavoitehintaisille rahavarauksille

rckopteri committed:
Lisää mahdollisuus tallentaa tavoitehintaiset

rahavaraukset kuluvalle vuodelle ja urakan tuleville vuosille

rckopteri committed:
Yksikkötestit tavoitehintaisille rahavarauksille

rckopteri committed:
Korjaa hidas app-staten alustus

rckopteri committed:
Tallenna tavoitehintaiset rahavaraukset aina

normaalin kalenteriajan mukaan eli ensimmäiselle vuodelle 3 kuukautta ja toiselle vuodelle loput 9

solita-staskila committed:
Näytä nolla-arvot, Järjestä rahavaraukset aakkosilla

solita-staskila committed:
Integroi rahavaraukset  tuotekoht laskutusyhteenvetoon

rckopteri committed:
Varmista, että rahavaraukset eivät luo

tyhjää riviä kustannusten seurantaan UI:lle

rckopteri committed:
Piilota turhia DEBUG laatikoita

rckopteri committed:
Varmista, että mahdolliset desimaalitäsmäyssummat

menevät hoitovuoden viimeiselle kuukaudelle

rckopteri committed:
Merge branch 'rahavaraukset-master' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

# Conflicts:
#	src/cljs/harja/views/urakka/kulut/kulut.cljs

rckopteri committed:
Poista erillishankinnat tehtäväryhmä

kulu lomakkeelta, kun kulutyypiksi valitaan hankintakulu

rckopteri committed:
Lisää muut kulut ja muuttuneet johto ja hallintkokulut

Kustannusten Seurannan exceliin

rckopteri committed:
Jäsennä Kustannusten Seurannan Excel ja välikatselmusnapit

Oikean kokoisiksi ja oikeaan kohtaan.

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1125__.sql

rckopteri committed:
Korjaa vuosikäsittely insertoidulle tavoitehintaiselle rahavaraukselle

rckopteri committed:
Estä duplikaatti tehtäväryhmien listaus rahavarauksille

rckopteri committed:
Korjaa kulu-lomakkeen validointi

rckopteri committed:
Pieni refaktorointi ui:lle

rckopteri committed:
Piilota vanhan tyyppiset rahavaraukset kustiksesta

rckopteri committed:
Lisää urakoille mahdolliset puuttuvat rahavaraukset

vanhojen rahavaraustehtävien perusteella

rckopteri committed:
Siirrä tavoitehinnan ulkopuoliset rahavaraukset oikeaan kohtaan

solita-staskila committed:
Näytä vain urakalle asetetut rahavaraukset työmaakokous

solita-staskila committed:
Näytä vain urakalle asetetut rahavaraukset tuotekoht

solita-staskila committed:
Linkitä rivit oikein toimenpideinstansseille

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1127__.sql

rckopteri committed:
Muokkaa tilaajan rahavaraukset eli tavoitehinnan ulkopuoliset rahavaraukset

toimimaan ilman gridiä kustannusten suunnittelussa.

rckopteri committed:
Mahdollista muokkausgridin otsikon muokkaus

rckopteri committed:
Päivitä testit vastaamaan muuttunutta datarakennetta

solita-staskila committed:
Lisää yhteensä- sarake rahavarausten alle työmaakokous

solita-staskila committed:
Tarkenna taulukon otsikko

solita-staskila committed:
Laske muut tavoitehinta kulut, työmaakokous

solita-staskila committed:
Näytä tavoitehintaiset muut kulut, työmaakokous

solita-staskila committed:
Laske muut ei tavoitehintaiset kulut, työmaakokous

solita-staskila committed:
Näytä ei tavoitehintaiset muut kulut, työmaakokous

solita-staskila committed:
Varmista että tavoitehintaiset kulut haetaan tavoitehintaisina

solita-staskila committed:
Poista äkilliset, vahingot yksittäiset muuttujat turhina, tuotekoht

solita-staskila committed:
Poista tilaajan rahavaraus termit ja laskennat turhina, tuotekoht

solita-staskila committed:
Korvaa yksilöivä tunniste rahavaraus id:llä, tuotekoht.

solita-staskila committed:
Poista tavoitehintaan kuulumattomat rivit hankinnoista, tuotekohtainen

solita-staskila committed:
Näytä muut kulut taulukko, tuotekohtainen

solita-staskila committed:
Laske muut kulut jokaisen instanssin alle, tuotekohtainen

solita-staskila committed:
Poista tilaajan rahavaraus termi tuotekohtaisesta

solita-staskila committed:
Laske muut tav. ulkopuoliset laskutukseen, työmaakokous

solita-staskila committed:
Poista ylimääräinen välitaulukko, laske tav. muut kulut mukaan, työmaakokous

solita-staskila committed:
Laske kaikki yht. oikein, työmaakokous

solita-staskila committed:
Näytä muut kulut aina, tuotekoht.

solita-staskila committed:
Näytä rahavaraukset ja muut kulut kaikille instansseille, tuotekoht

solita-staskila committed:
Laske tav. hintaiset, ja ei tav. hintaiset muut kulut yhteensä, tuotekohtainen

solita-staskila committed:
Laske tavoitehintaan kuuluvat oikein, tuotekohtainen

solita-staskila committed:
Refaktoroi taulukkojen koostaminen tuotekohtaiseen yhteenvetoon

solita-staskila committed:
Merge branch 'rahavaraukset-laskutus' into rahavaraukset-master

rckopteri commented on tietokanta/src/main/resources/db/migration/R__Rahavaraukset.sql:
```diff
@@ -0,0 +1,135 @@
+-- Rahavaraus id:n lisäys ja populointi -- SELECT populoi_rahavaraus_idt();
+-- Palauttaa päivitetyt rivit yhteenlaskettuna
+CREATE OR REPLACE FUNCTION populoi_rahavaraus_idt()
```
Ota kulu_kohdistus ja kustannusarvioitutyö ja toteutuneet_kustannukset taulut johonkin talteen ennen tätä migraatiota, niin ei vahingossakaan menetetä mitään.

rckopteri committed:
Lisää järjestys rahavarausten tehtäväryhmille

rckopteri committed:
UI korjauksia kululomakkeelle

rckopteri committed:
Poista vanha kululomake ja korjaa aikavalinnat kululistauksesta

rckopteri committed:
Nollaa päivämäärävalinnat, kun muutetaan muita filttereitä kulut listauksessa

rckopteri committed:
Merge branch 'rahavaraukset-master' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

rckopteri committed:
Päivitä migraatiotiedostojen nimet

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

rckopteri committed:
Lisätään kustannusarvioitu_tyo taulun rahavarauksille

myös oikea osio ja korjataan puuttuvien rahavarausten kopionti

rckopteri committed:
Päivitä rahavaraukset uusimpaan arvioon siitä mitä niiden tulee olla

rckopteri committed:
Lisää korjauksia rahavarausten tietokantatilanteeseen

rckopteri committed:
Järjestä rahavaraukset hallintaan id järjestykseen

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1128__.sql

rckopteri committed:
Korjaa rahavaraukset tietokantaan

rckopteri committed:
Näytä latausindikaattori, kun kuluja haetaan

rckopteri committed:
Näytä urakkakohtainen rahavarausnimi, jos se on asetettu

rckopteri committed:
Hae urakkakohtainen rahavarausnimi aina kululle

rckopteri committed:
Korjaa eräpäivä muille, kuin alkaville urakoille

rckopteri committed:
Varmista, että rahavaraus_urakka tauluun ei voi tulla duplikaattia

maaritlaine committed:
Merge branch 'develop' into rahavaraukset-master

solita-staskila committed:
Tavoitehintainen pois rivihausta

solita-staskila committed:
Poista turha deprikoitunut migraatio

solita-staskila committed:
Tallenna rahavaraukset aina tavoitehintaisena

rckopteri committed:
Korjaa kulun tallennus

rckopteri committed:
Päivitä tehtävät rahavarauksia varten

rckopteri committed:
Merge branch 'rahavaraukset-master' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

solita-staskila committed:
Korjaa tavoitehintaiset hankinnat laskutukseen, kulu lomakkeeseen

solita-staskila committed:
Alusta hankintojen muuttujat laskutukseen

solita-staskila committed:
Merge branch 'rahavaraukset-master' into rahavaraukset-laskutus

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1129__.sql
#	tietokanta/src/main/resources/db/migration/V1_1130__.sql

solita-staskila committed:
Korjaa päällysteiden paikkaus kytkökset kulu_kohdistukseen sekä tehtävä tauluun

rckopteri committed:
Poistetaan toimenpideinstanssin pakollisuus kustannusarvioitu_tyo taulun rahavarauksilta.

rckopteri committed:
Korjataan rahavarausten nimiä tehtäväryhmiä ja tehtäviä

rckopteri committed:
Suunnitellut tehtävät - formatoi luvut pilkullisiksi ja mahdollista pistellä tallentaminen

solita-staskila committed:
Looppaa vain teiden-hoito, looppaa vain ppaikkaus tehtäväryhmä

solita-staskila committed:
Siivoa kytkösmigraatiofunktio

solita-staskila committed:
Korjaa kulujen näyttö tuotekohtaiseen, korjaa kannustin id

solita-staskila committed:
Korjaa duplikaatti hoitokausi rivit (huoh)

solita-staskila committed:
Formatointi

solita-staskila committed:
Nosta migraatiota

solita-staskila committed:
Merge branch 'rahavaraukset-master' into rahavaraukset-laskutus

solita-staskila committed:
Laske migraatio

solita-staskila committed:
Nosta rahavarausten migraatiot

solita-staskila committed:
Merge branch 'develop' into rahavaraukset-master

solita-staskila committed:
Laske rahavarausten migraatiot

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	test/clj/harja/palvelin/palvelut/hallinta/urakoiden_rahavaraukset_test.clj
#	tietokanta/src/main/resources/db/migration/V1_1131__.sql
#	tietokanta/src/main/resources/db/migration/V1_1132__.sql

rckopteri committed:
Kohdista kaikki rahavaraustietokantamuutokset yhteen tiedostoon

rckopteri committed:
Näytä rahavarauksen default nimi, vaikka urakkakohtainen nimi olisi asetettu tyhjäksi stringiksi

rckopteri committed:
Näytä kustannusten seurannassa rahavarauksen nimi

vaikka urakkakohtainen nimi olisi asetettu tyhjäksi stringiksi

rckopteri committed:
Järjestä rahavaraukset jarjestys-sarakkeen mukaisesti

solita-staskila committed:
Korjaa tehtäväryhmän nimi täsmäämään uutta develop branchia

rckopteri committed:
Mahdollista tavoitehintaisten rahavarausten vahvistaminen

rckopteri committed:
Laita tulevat hoitovuodet defaulttina päälle

rckopteri committed:
Korjaa rahavarausten formatoinnit

solita-staskila committed:
Päivitä maksuerät sampoon päivitetyille toimenpideinstansseille

rckopteri committed:
Kululomakkeen ulkonäköparannuksia

rckopteri committed:
Merge branch 'rahavaraukset-master' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

rckopteri committed:
Siirretään enumin käsittely omaan transaktioon

rckopteri committed:
Järjestä rahavaraukset tietokantaan oikein

rckopteri committed:
Poista käyttämätön koodi

rckopteri committed:
Järjestä rahavaraukset kustannusten seurantaan oikein

rckopteri committed:
Korjauksia kulujen ui:hin ja kohdistuksen poiston korjaus

rckopteri committed:
Muuta vielä pvm kentän ulkoasu enemmän muiden input kenttien tyyppiseksi

rckopteri committed:
Päivitä testit vastaamaan toteutusta

solita-staskila committed:
Poista turhat requiret

rckopteri committed:
Poista rahavarausten osuus hankintakustannusten yhteenvedosta

rckopteri committed:
Nimeä Hankintakustannukset -> Suunnitellut hankinnat

Nimi vastaa paremmin tarjous dokumentteja ja kustannusten seurantaa

rckopteri committed:
Korjaa phantom testit, kun kululomake on muuttunut

rckopteri committed:
Ota varmuuskopiot muokattavista tauluista, ennen muokkausta

rckopteri committed:
Varmuuskopiotaulujen vuoksi TYPEä ei voi poistaa

solita-staskila committed:
Korjaa kovakoodattu arvo liikenne raporteissa

solita-staskila commented:
Hox tuotannossa **_Muut kustannukset yhteensä_** alle ei lasketa rahavarauksia, mutta nykyisessä speksissä lasketaan. Kaikki luvut täsmäävät muuten, mutta tämä rivi muuttuu.  Täytynee tarkistaa designereilta, että tämä on oikein.
<details>
  <summary>>>screenshot</summary>
  
  ![image](https://github.com/user-attachments/assets/157f74cc-1e00-4a41-805e-3fafbb4e6da3)

</details>

solita-staskila committed:
Korjaa yhteensä sarake rivi alimmaksi, työmaakokous

solita-staskila committed:
Laske ei tav. muut yht oikein, työmaakokous

solita-staskila committed:
Refaktoroi työmaak. UI uuden Figman speksin mukaiseksi 12.9

solita-staskila committed:
Poista ylimääräiseksi jäänyt hoidonjohto-valitaulukko

solita-staskila committed:
Refaktoroi raportin generointia, poista tarpeettomat funktiot

solita-staskila committed:
Laske hank, hoid. johto, tavoitehinnat rahavarauksineen oikein

solita-staskila commented:
Tuotannossa, lasketaan tällä hetkellä tuotekohtaisessa laskutusyhteenvedossa tavoitehintaiset kustannukset väärin, niistä puuttuu rahavaraukset. Tähän tulossa tämän branchin mukana korjaus.

solita-staskila committed:
Korjaa tiedostonimi

solita-staskila committed:
Siivoa turhia requireja

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	tietokanta/src/main/resources/db/migration/V1_1133__.sql
#	tietokanta/src/main/resources/db/migration/V1_1134__.sql

rckopteri committed:
Siirrä toteutuneet_kustannukset taulun funktiot omaan r migraatioon

Sinne siirtyi update, delete triggerit sekä funktio, joka ajaa datan tauluun päiväkohtaisesti

rckopteri committed:
Päivitä migraatioon tehtäväryhmien nimet ja kommentoi pois taulujen kopiointi

rckopteri committed:
Validoi lomake kulukohdistuksen poiston jälkeen.

solita-staskila committed:
Merge branch 'rahavaraukset-master' into rahavaraukset-laskutus

solita-staskila commented on src/cljs/harja/views/kanavat/urakka/liikenne.cljs:
```diff
@@ -605,7 +606,7 @@
    (if rivimaara-ylittynyt
      [:div.kayttajia-ei-loydy.fontti-14
-      [yleiset/keltainen-vihjelaatikko [:div "Tulokset rajattu 10 000 tapahtumaan. Tarkenna hakua."] :info]
+      [yleiset/keltainen-vihjelaatikko [:div "Tulokset rajattu " (fmt/formatoi-numero-tuhansittain lt/+rajoita-tapahtumien-maara+) " tapahtumaan. Tarkenna hakua."] :info]
```
Tainnut tulla pushi väärään haaraan, ei liity rahavarauksiin mitenkään. Tässä on korjattu kovakoodattu arvo, mutta saa olla kyllä täälläkin.

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

rckopteri committed:
Merkitse rahavaraukset vahvistetuiksi jos ne voidaan vahvistaa.

Hankintakustannukset on oltava vahvistettuina ja rahavarauksissa ei saa olla viitettä Muut tavoitehintaiset rahavaraukset rahavaraukseen.

solita-staskila committed:
Lisää urakkakohtainen rahavarausnimi työmaakokous laskutukseen

solita-staskila committed:
Lisää urakkakohtainen rahavarausnimi tuotekoht laskutukseen

solita-staskila committed:
Aseta urakkanimi NULL jos annetaan tyhjä string

solita-staskila committed:
Mäppää rahavaraukset testeihin, tavoitehinta-toimii työmaaraportti

rckopteri committed:
Poista turhia logeja

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

# Conflicts:
#	src/clj/harja/palvelin/raportointi/excel.clj

solita-staskila committed:
Pura tuetokoht oikein testeihin, testaa talvihoito

solita-staskila committed:
Testaa liikenneymp. hoito, tuotekoht.

solita-staskila committed:
Testaa soratien hoito, tuotekoht.

solita-staskila committed:
Testaa päällyste, tuotekoht.

solita-staskila committed:
Testaa Mhu-Hju hoidon johto, tuotekoht.

solita-staskila committed:
Testaa Mhu ylläpito, tuotekoht.

solita-staskila committed:
Testaa Mhu korvausinvestointi, tuotekoht.

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

rckopteri committed:
Poista muilta kuluilta kustannusten seurannasta suunniteltujen summien näyttäminen.

Niille ei suunnitella mitään, joten turha näyttää

solita-staskila committed:
Älä laske rarhavarauksia Ylläpito riville HOX alkup. toteutus muuttuu

solita-staskila committed:
Rahavaraukset pois hankitakuluista työmaakok. HOX alkup. toteutus muuttuu

rckopteri committed:
Merge branch 'develop' into rahavaraukset-master

solita-staskila committed:
Korjaa rahavarausten järjestys laskutusyhteenvedolla

solita-staskila committed:
Korjaa default koontilaskun kuukausi kulukijaukseen

rckopteri committed:
Poista kustiksen suunnitelluista hankinnoista rahavaraukset

rckopteri committed:
Salli nil arvon tallentaminen rahavaraukselle kustannusten suunnittelussa

solita-staskila committed:
Korjaa eräpäivä, jos nyk aika on yli urakan loppupvm

solita-staskila committed:
Lisää info & lukitusikonit alasvetoon jos hoitovuodelle on päätöksiä

solita-staskila committed:
Siirrä margin oikeaan elementtiin

solita-staskila committed:
Kytke tuotekohtaisen rahavaraukset instansseihin oikein

solita-staskila committed:
Heitä muut kulut tuotekohtaisesta toistaiseksi kuikkaan

solita-staskila committed:
Kohdista T3 & muut rahavaraukset tuotekohtaiseen oikein

solita-staskila committed:
Merge branch 'rahavaraukset-master' into rahavaraukset-laskutus

solita-staskila committed:
Merge branch 'develop' into rahavaraukset-master

solita-staskila committed:
Korjaa tuotekoht testit uusiin muutoksiin

solita-staskila committed:
Korjaa rahavarausten järjestys budjettisuun. testiin

solita-staskila commented on src/clj/harja/palvelin/raportointi/excel.clj:
```diff
@@ -527,7 +435,7 @@
         (fn [rivi-nro rivi]
           ;; Lisää väliotsikot mikäli nämä puuttuvat 
-          (let [lisatty-otsikko (when (:otsikko rivi)
+          (let [lisatty-otsikko (when (and (:otsikko rivi) lisaa-excel-valiotsikot)
```
Tämä (lisaa-excel-valiotsikot) pitää heittää kuikkaan, en tiedä miksi tullut tänne, developpi  haarassa tämä on juuri  poistettu.

rckopteri replied:
Varmaan jossain konfliktin ratkomisessa on tullut virhe. Kokeilin korjata.

samnes commented on src/clj/harja/palvelin/palvelut/budjettisuunnittelu.clj:
```diff
@@ -939,6 +985,136 @@
+                                                                            :sopimus_id sopimus-id})
+                  ;; Jokaisella kustannusarvoitu_tyo -rivillä pitää olla toimenpideinstanssi.
+                  ;; Tällä hetkelä ei tiedetä, että saako rahavaraukset oman toimenpideinstanssin, vai otetaanko joku vain.
+                  ;; Joten tässä vähän keskeneräinen toteutus, jossa toimenpideinstansseista otetaan vain ensimmäinen.
```
Pitääkö kommentti yhä paikkaansa? Tuliko asialle selkeys?

rckopteri replied:
Päivitin kommentin. Rahavarauksille ei kuulu toimenpideinstanssia, mutta en lähtenyt poistamaan sen pakollisuutta, koska se vaikuttaa sekä tietokantaan, että moniin hakuihin kustiksessa.

rckopteri committed:
Päivitä kustannusten kirjaus cypress testi vastaamaan uutta toteutusta

maaritlaine commented on tietokanta/src/main/resources/db/migration/R__Laskutusyhteenveto_mhu.sql:
```diff
@@ -644,549 +654,646 @@ BEGIN
-                                                       AND m.toimenpideinstanssi = tpi.id
-                                                       AND tpk2.koodi = '23150'
-                                                     limit 1);
+    johto_ja_hallintakorvaus_toimenpideinstanssi_id := (
```
Tässä on ollut alunperinkin ylimääräistä monimutkaisuutta. On turha hakea tietoja toimenpide.taso = 2 kautta. Käytetään suoraan tason 3 toimenpidekoodia (23151), eikä sotketa mukaan maksuerätaulua tarpeettomasti. Limit 1 myös pois. Jos rivejä palautuu liikaa, on hyväkin saada virhe.

SELECT tpi.id AS id
FROM toimenpideinstanssi tpi
         JOIN toimenpide tp ON tpi.toimenpide = tp.id
WHERE tpi.urakka = ur
  AND tp.koodi = '23151';

maaritlaine commented on tietokanta/src/main/resources/db/migration/R__Laskutusyhteenveto_mhu.sql:
```diff
@@ -644,549 +654,646 @@ BEGIN
-                 ORDER BY m.numero asc
+    FOR t IN 
+        SELECT 
+          tpk2.nimi AS nimi, 
```
Tässäkin on ylimääräistä ja jatkossa lisää tason 2 tuotekoodia. Ehkä jätetään tässä vaiheessa ennalleen ja refaktoroidaan toisessa yhteydessä.

solita-staskila replied:
Kirjasin nämä tiksulle, tehdään myöhemmin

rckopteri committed:
Refaktoroi kustannussuunnitelman tiedot ui controlleria

rckopteri committed:
Tarkenna cy testejä varten elementtien nimet ja id:t

rckopteri committed:
Päivitä cy testit vastaamaan muuttunutta kustannusten suunnittelua

rckopteri committed:
Refaktoroidaan katselmointikommenttien perusteella

solita-staskila committed:
Hae tuotekohtaiseen indeksikorjattu taveoitehinta

rckopteri committed:
Paranna johto ja hallintokorvausten cy testejä

rckopteri committed:
Merge branch 'rahavaraukset-master' of https://github.com/finnishtransportagency/harja into rahavaraukset-master

rckopteri committed:
Lisää formatointi koontilaskun kuukauden valintakomponenttiin

rckopteri committed:
Laita kululomakkeen testaus toimimaan

rckopteri committed:
Mahdollista vielä negaatiivisten kulujen syöttö ja varmista se cy testillä

solita-staskila commented:
Chekkaatko tämän @rckopteri 

Tulee compile warning (lein cleanin jälkeen kun käynnistät frontin):

#### **Use of undeclared Var harja.tiedot.urakka.suunnittelu.mhu-kustannussuunnitelma/rahavarausten-rajapinta**

 ```clojure
  628  
  629      (e! (tuck-apurit/->MuutaTila [:gridit :rahavaraukset :grid] g))
  630  
  631      (grid/rajapinta-grid-yhdistaminen! g
  632        t/rahavarausten-rajapinta
             ^---
  633        (t/rahavarausten-dr)
  634        {[::t/otsikko] {:rajapinta :rahavaraukset-otsikot
  635                        :solun-polun-pituus 1
  636                        :jarjestys [^{:nimi :mapit} [:nimi :maara :yhteensa :indeksikorjattu]]
  637                        :datan-kasittely (fn [otsikot]
```

####  **Use of undeclared Var harja.tiedot.urakka.suunnittelu.mhu-kustannussuunnitelma/rahavarausten-dr**

```clojure
  628  
  629      (e! (tuck-apurit/->MuutaTila [:gridit :rahavaraukset :grid] g))
  630  
  631      (grid/rajapinta-grid-yhdistaminen! g
  632        t/rahavarausten-rajapinta
  633        (t/rahavarausten-dr)
              ^---
  634        {[::t/otsikko] {:rajapinta :rahavaraukset-otsikot
  635                        :solun-polun-pituus 1
  636                        :jarjestys [^{:nimi :mapit} [:nimi :maara :yhteensa :indeksikorjattu]]
  637                        :datan-kasittely (fn [otsikot]
  638                                           (mapv (fn [otsikko]
```

solita-staskila committed:
Poista turhat parametrit

solita-staskila committed:
Disabloi haku, jos haku on jo käynnissä (korjaa bugeja)

solita-staskila committed:
Näytä hyrrä myös silloin kun kohdistusta ladataan

solita-staskila commented:
> Ota kulu_kohdistus ja kustannusarvioitutyö ja toteutuneet_kustannukset taulut johonkin talteen ennen tätä migraatiota, niin ei vahingossakaan menetetä mitään.

Onko tehty??
https://github.com/finnishtransportagency/harja/pull/3499#discussion_r1723074299

rckopteri committed:
Poista viimeiset jäämät rahavarausgridistä kustiksesta.

rckopteri committed:
Estä null pointer virhe cy testeissä
